### PR TITLE
Migrate from Zalo Login API v3  to V4

### PR DESCRIPTION
**Pull request by binhuq**
Zalo Login API v3 had already deprecated https://developers.zalo.me/docs/api/developer-notification/thong-bao-ngung-ho-tro-hoan-toan-dich-vu-zalo-login-v3-post-6905?campaign=DEVELOPER_BANNER_CAMPAIGN. Migrate to v4 so zalo provider can working again.

This is my first pull request. Please give feedback if have some problems.

Sorry for my bad English, English is not my first language.

### CODE REVIEW

File: src/Zalo/Provider.php [hunk 1]
```diff
@@ -8,6 +8,9 @@
 
 class Provider extends AbstractProvider
 {
+    /**
+     * Unique Provider Identifier.
+     */
```
Comment (atymic) [inline src/Zalo/Provider.php:13]: No need for this comment

File: src/Zalo/Provider.php [hunk 2]
```diff
@@ -20,52 +23,64 @@ class Provider extends AbstractProvider
...
-        $response = $this->getHttpClient()->get($this->getTokenUrl(), [
-            RequestOptions::HEADERS => ['Accept' => 'application/json'],
-            RequestOptions::QUERY   => $this->getTokenFields($code),
+        $response = $this->getHttpClient()->post($this->getTokenUrl(), [
+            RequestOptions::HEADERS     => $this->getTokenHeaders($code),
+            RequestOptions::FORM_PARAMS => $this->getTokenFields($code),
         ]);
 
-        return json_decode((string) $response->getBody(), true);
+        return json_decode($response->getBody(), true);
```
Comment (atymic) [inline src/Zalo/Provider.php:51]: String cast here please

File: src/Zalo/Provider.php [hunk 3]
```diff
@@ -20,52 +23,64 @@ class Provider extends AbstractProvider
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getTokenHeaders($code)
+    {
+        return [
+            'Content-Type' => 'application/x-www-form-urlencoded',
```
Comment (atymic) [inline src/Zalo/Provider.php:60]: No need to set this header when using the form params request option

File: src/Zalo/Provider.php [hunk 4]
```diff
@@ -20,52 +23,64 @@ class Provider extends AbstractProvider
...
+            'secret_key'   => $this->clientSecret,
```
Comment (atymic) [inline src/Zalo/Provider.php:61]: As this is the only header can you just set it in the get access token response method?

File: src/Zalo/Provider.php [hunk 5]
```diff
@@ -20,52 +23,64 @@ class Provider extends AbstractProvider
...
+        return 'https://oauth.zaloapp.com/v4/access_token';
     }
 
     /**
      * {@inheritdoc}
      */
     protected function getUserByToken($token)
     {
-        $response = $this->getHttpClient()->get('https://graph.zalo.me/v2.0/me?access_token='.$token.'&fields=id,birthday,name,gender,picture');
+        $response = $this->getHttpClient()->get('graph.zalo.me/v2.0/me', [
```
Comment (atymic) [inline src/Zalo/Provider.php:78]:
```suggestion
        $response = $this->getHttpClient()->get('https://graph.zalo.me/v2.0/me', [
```

File: src/Zalo/Provider.php [hunk 6]
```diff
@@ -20,52 +23,64 @@ class Provider extends AbstractProvider
...
+            RequestOptions::HEADERS => ['access_token' => $token],
+            RequestOptions::QUERY   => ['fields' => 'id,error,message,name,picture'],
+        ]);
 
-        return json_decode((string) $response->getBody(), true);
```
Comment (atymic) [inline src/Zalo/Provider.php:68]: Leave string cast please

**Review by atymic — CHANGES_REQUESTED** [review]

**Comment by atymic** [comment]
Do you have a link to english docs? I cant get translate to work on the viet ones.

**Comment by binhuq** [comment]
> Do you have a link to english docs? I cant get translate to work on the viet ones.

Here is english doc. If you don't see english please scroll to bottom and chose English
https://developers.zalo.me/docs/api/social-api-4

**Review by atymic — APPROVED** [review]

**Pull request merged**